### PR TITLE
ZeroOrMore: avoid instantiating builder when no match

### PR DESCRIPTION
**Pull request by yanns**
Fix https://github.com/sirthias/parboiled2/issues/319

### CODE REVIEW

**Comment by yanns** [comment]
I've broken `scalaParser/testOnly scalaparser.SnippetSpec`. Let me check.

**Comment by yanns** [comment]
Fix: https://github.com/sirthias/parboiled2/pull/386/commits/8de614123c93438193c4ea873adea48b05c8fae5

**Comment by yanns** [comment]
@jrudolph when you find time, could you have a look at this?

**Comment by sirthias** [comment]
Thank you, @yanns, for finding and reporting the problem and proposing this fix!
Your solution looks ok to me but there is one problem that might justify another (and maybe even simpler) approach:

The problem is that where `zeroOrMore` was previously rendering its inner op (`op.render(wrapped)`) only once it now rendered it twice because you unrolled the first iteration.
If the inner op is simple and small this is completely fine.

It could be though that the inner op is large (even really large).
Rendering it twice could end up greatly increasing code side with all its drawbacks (decreasing locality, inlining horizons, etc.).

So it might be better to not unroll the first iteration but to simply postpone the instantiation of the builder until the first actual match. This costs us another `null` comparison in the loop which I'd consider acceptable.

WDYT?

**Comment by yanns** [comment]
That's a good point. I'll have a look at this.

**Comment by yanns** [comment]
@sirthias you can have a look at the last version that avoids the double rendering of the inner op.

File: parboiled-core/src/main/scala-2/org/parboiled2/support/OpTreeContext.scala [hunk 1]
```diff
@@ -748,20 +749,42 @@ trait OpTreeContext[OpTreeCtx <: ParserMacros.ParserContext] {
...
+    def withLazyBuilder: Collector = this
+  }
 
   lazy val rule0Collector = {
     val unit = q"()"
     new Collector(unit, unit, q"true", unit, unit)
   }
 
+  lazy val lazyRule1Collector = new Collector(
+    valBuilder = q"var builder: scala.collection.immutable.VectorBuilder[Any] = null",
```
Comment (yanns) [inline parboiled-core/src/main/scala-2/org/parboiled2/support/OpTreeContext.scala:762]: Maybe I should rename `valBuilder` to `declareBuilder`?

Comment (sirthias) [reply parboiled-core/src/main/scala-2/org/parboiled2/support/OpTreeContext.scala:762]: If you want to.
But I think it's fine as it is as well.

**Comment by yanns** [comment]
yes totally!

**Comment by yanns** [comment]
Would it be possible to cut a release with this "feature"? That would be awesome!

**Comment by sirthias** [comment]
Ok, I just published `2.4.1` to sonatype.
Should appear on maven central soon.
Thank you again for contributing this fix, Yann!

**Pull request merged**